# Librerías

In [ ]:
#!pip install pysqlite3
#!pip install scikit-lear
!pip install -q google-generativeai #Para API de Gemini
!pip install requests #Libreria no totalmente necesaria pero recomenda en caso de peticiones HTTP
!pip install numpy==1.23.5
!pip install scikit-surprise

In [ ]:
# Manejo de Datos
import numpy as np
import pandas as pd
import sqlite3 as sql
import os
import sys
import datetime

# Visualización
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.graph_objs as go
import plotly.express as px
from plotly.subplots import make_subplots

# Estadísticas y Pruebas
import scipy.stats as stats
from scipy.stats import gaussian_kde

# Procesamiento de Datos
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.decomposition import PCA

# Modelado y Algoritmos
from sklearn import neighbors
import joblib
#from surprise import Reader, Dataset
from surprise.model_selection import cross_validate, GridSearchCV
from surprise import KNNBasic, KNNWithMeans, KNNWithZScore, KNNBaseline
from surprise.model_selection import train_test_split

# Interactividad
from ipywidgets import interact

# Google Colab
from google.colab import drive
from google.colab import files

# Otros
from collections import Counter

#Librerias de la API fe Gemini
import google.generativeai as genai #Para API de Gemini
import requests # Se usa para hacer solicitudes HTTP. Permite enviar peticiones POST a las APIs para traducción y generación de texto

In [ ]:
from surprise import Reader, Dataset #Libreria para modelos de recomendación

# Conectar con google drive


In [ ]:
drive.flush_and_unmount() #Linea en caso de tener que desconectar el drive por algún tipo de falla

Drive not mounted, so nothing to flush and unmount.


In [ ]:
#drive.flush_and_unmount()  #Linea en caso de tener que desconectar el drive por algún tipo de falla
drive.mount('/content/drive') #Linea para conectar al drive

Mounted at /content/drive


In [ ]:
path="/content/drive/MyDrive/analitica 3/sistemas_recomendacion" ### ruta del repositorio en drive
os.chdir(path) ### volver la carpeta del repositorio directorio de trabajo
sys.path.append(path) ### agregarla al path, poder leer archivos de funciones propios como paquetes

In [ ]:
import a_funciones as fn #Importar el documento de funciones para hacer uso de estas

# Base de datos

In [ ]:
conn = sql.connect('/content/drive/MyDrive/analitica 3/sistemas_recomendacion/data/db_movies3') #Crear la conexión con la base de datos
cur = conn.cursor() #Creacion del cursos para realizar consultas dentro del mismo SQL

In [ ]:
# Creación de cursor para  ejecutar consultas en la base de datos
# Visualizar las tablas contenidas en la base de datos
cur.execute("SELECT name FROM sqlite_master where type='table'")
cur.fetchall()

[('movies_final',),
 ('ratings_final',),
 ('df_explorado',),
 ('df_terminado',),
 ('df_catalogo',),
 ('reco',)]

#Modelos

In [ ]:
df_modelos=joblib.load('/content/drive/MyDrive/analitica 3/sistemas_recomendacion/salidas/df_modelos.joblib')   # Carga el DataFrame de modelos desde un archivo .joblib

In [ ]:
df_catalogo = pd.read_sql("SELECT * from df_catalogo",conn) #Guardar Dataframe en variable

## Sistema de recomendación basado en contenido KNN

In [ ]:
#Se reinicia el índice de ambos DataFrames para que esten alineados y no den problemas
df_modelos = df_modelos.reset_index(drop=True)
df_catalogo = df_catalogo.reset_index(drop=True)

In [ ]:
#Se crea el modelo KNN con 10 vecinos y usando la distancia euclidiana
model=neighbors.NearestNeighbors(n_neighbors=11,metric='euclidean')
model.fit(df_modelos)
dist,idlist=model.kneighbors(df_modelos) # se calculan las distancias y los indices de los 10 vecinos más cercano

In [ ]:
#Convertimos las matrices dist (distancias) e idlist (índices de vecinos) en dataframes de pandas para facilitar su uso posterior
distancias=pd.DataFrame(dist)
id_list=pd.DataFrame(idlist)

**Modelo con película en específico**

In [ ]:
movie_list_name = []  # Lista vacía para almacenar títulos recomendados
movie_name='Avengers: Infinity War - Part I (2018)' # Película ejemplo
movie_id = df_catalogo[df_catalogo['title'] == movie_name].index ### extraer el indice de la película
movie_id = movie_id[0] ## si encuentra varios solo guarde uno


# Guardar la lista de peliculas recomendadas
for newid in idlist[movie_id]:
    movie_list_name.append(df_catalogo.loc[newid].title) ### agrega el nombre de cada una de los id recomendados

# Mostrar películas mas parecidas nuestro ejemplo específico
movie_list_name


['Avengers: Infinity War - Part I (2018)',
 'Guardians of the Galaxy 2 (2017)',
 'Guardians of the Galaxy (2014)',
 'Thor: Ragnarok (2017)',
 'Serenity (2005)',
 'X-Men: Days of Future Past (2014)',
 'Iron Man (2008)',
 'Black Panther (2017)',
 'Ant-Man (2015)',
 'Doctor Strange (2016)',
 'X-Men (2000)']

**Modelo general para cualquier película**

In [ ]:
# Modelo general como función que toma como entrada el título de las películas
def movie_recomender(movie_name=list(df_catalogo['title'].value_counts().index)):
    movie_list_name = []
    movie_id = df_catalogo[df_catalogo['title'] == movie_name].index ### extraer el indice del libro
    movie_id=movie_id[0] ## si encuentra varios solo guarde uno

# Iterar sobre los 10 vecinos de esa película para evitar titulos duplicados
    for newid in idlist[movie_id]:
        movie_list_name.append(df_catalogo.loc[newid].title) ### agrega el nombre de cada una de los id recomendados
    return movie_list_name

print(interact(movie_recomender)) # Activamos el selector interactivo para elegir una película y obtener recomendaciones

Al encontrar las películas más similares en función de sus características, el modelo identifica los vecinos más cercanos (por ejemplo, las 10 películas más similares a Toy Story (1995)), mostrando títulos que comparten similitudes como el genero, años, calificaciones, e incluso aunque no es una variable que estemos considerando nos da similitud en cuanto al estilo de pelicula. Esto demuestra que el enfoque basado en KNN es efectivo para ofrecer sugerencias coherentes, especialmente en catálogos donde la similitud de contenido entre películas mejora la experiencia del usuario.

##Sistema de recomendación basado en el usuario KNN

In [ ]:
## seleccionar usuarios para recomendaciones
usuarios=pd.read_sql('select distinct (user_id) as user_id from ratings_final',conn)

In [ ]:
# Función para recomendar películas a un usuario basado en su perfil de gustos usando k-vecinos más cercanos
def recomendar(user_id=list(usuarios['user_id'].value_counts().index)):

    ###seleccionar solo los ratings del usuario seleccionado
    ratings=pd.read_sql('select *from ratings_final where user_id=:user',conn, params={'user':user_id})

    ###convertir ratings del usuario a array
    l_movies_r=ratings['movie_id'].to_numpy()

    ###agregar la columna de movie_id y titulo de la pelicula a dummie para filtrar y mostrar nombre
    df_modelos[['movie_id','title']]=df_catalogo[['movie_id','title']]

    ### filtrar las peliculas calificadas por el usuario
    movies_r=df_modelos[df_modelos['movie_id'].isin(l_movies_r)]

    ## eliminar columna nombre e movie_id
    movies_r=movies_r.drop(columns=['movie_id','title'])
    movies_r["indice"]=1 ### para usar group by y que quede en formato pandas tabla de centroide
    ##centroide o perfil del usuario
    centroide=movies_r.groupby("indice").mean()


    ### filtrar peliculas no vistas
    movies_nr=df_modelos[~df_modelos['movie_id'].isin(l_movies_r)]
    ## eliminbar nombre y movie_id
    movies_nr=movies_nr.drop(columns=['movie_id','title'])

    ### entrenar modelo
    model=neighbors.NearestNeighbors(n_neighbors=11, metric='cosine')
    model.fit(movies_nr)
    dist, idlist = model.kneighbors(centroide)

    ids=idlist[0] ### queda en un array anidado, para sacarlo
    recomend_b=df_catalogo.loc[ids][['title','movie_id']]
    leidos=df_catalogo[df_catalogo['movie_id'].isin(l_movies_r)][['title','movie_id']]

    return recomend_b


recomendar(550) #Ejecutar la funcion con user_id (usuario) de prueba

title  movie_id
1444                 Big Sleep, The (1946)      1284
1028        Star Trek Into Darkness (2013)    102445
1018                           Thor (2011)     86332
1443                          Akira (1988)      1274
938      Once Upon a Time in Mexico (2003)      6709
63              Clockwork Orange, A (1971)      1206
58                       Abyss, The (1989)      1127
1013               Expendables, The (2010)     79695
14                          Ed Wood (1994)       235
381              Once Were Warriors (1994)       290
197   Road Warrior, The (Mad Max 2) (1981)      3703

In [ ]:
print(interact(recomendar)) # Crea un widget interactivo para seleccionar el usuario y mostrar recomendaciones en tiempo real

Este sistema demuestra ser una herramienta útil para ofrecer recomendaciones automáticas y personalizadas en tiempo real, optimizando la experiencia del usuario sin requerir información explícita sobre sus gustos, lo único que se necesita es su historial de interacciones y usar el algoritmo de vecinos mas cercanos KNN para sugerir películas que el usuario aún no ha visto, basándose en sus preferencias previas

##Sistema de recomendación filtro colaborativo

In [ ]:
# Cargar toda la tabla 'ratings_final'
ratings=pd.read_sql('select * from ratings_final', conn)
ratings.head()

user_id  movie_id  rating
0        1         1     4.0
1        1         3     4.0
2        1         6     4.0
3        1        47     5.0
4        1        50     5.0

In [ ]:
reader = Reader(rating_scale=(0, 5)) # Define el rango de los ratings para el lector de datos de Surprise
data   = Dataset.load_from_df(ratings[['user_id','movie_id','rating']], reader) # Crea un conjunto de datos para Surprise a partir del dataframe de ratings

In [ ]:
models=[KNNBasic(),KNNWithMeans(),KNNWithZScore(),KNNBaseline()] # Lista de modelos KNN de Surprise para probar diferentes variantes
results = {} # Diccionario para almacenar resultados de los modelos

In [ ]:
model=models[1] # Elección de un modelo incial para la iteración
for model in models: # Iterar sobre cada modelo de la lista de modelos

    # Realizar validación cruzada (5 subdivisiones) calculando MAE y RMSE
    CV_scores = cross_validate(model, data, measures=["MAE","RMSE"], cv=5, n_jobs=-1)

    # Calculo del promedio de las metricas y guardar los resultados en el diccionario
    result = pd.DataFrame.from_dict(CV_scores).mean(axis=0).\
             rename({'test_mae':'MAE', 'test_rmse': 'RMSE'})
    results[str(model).split("algorithms.")[1].split("object ")[0]] = result


performance_df = pd.DataFrame.from_dict(results).T # Convertir diccionario de resultados en un dataframe
performance_df.sort_values(by='RMSE') # Ordenar los modelos en funcion de su RMSE de menor a mayor

MAE      RMSE  fit_time  test_time
knns.KNNBaseline     0.642137  0.840118  0.431126   5.287900
knns.KNNWithZScore   0.647144  0.851061  0.335988   2.119479
knns.KNNWithMeans    0.653280  0.854297  0.231783   2.342850
knns.KNNBasic        0.694662  0.904848  0.194383   1.674117

En los resultados obtenidos, el modelo KNNBaseline presenta el mejor desempeño, con el MAE más bajo (0.6421) y también el menor RMSE (0.8401), lo que indica una mayor precisión en la predicción de las calificaciones de los usuarios. Aunque no es el modelo más rápido en términos de tiempo de entrenamiento y prueba, la diferencia respecto a los demás es mínima y no representa un factor determinante.

In [ ]:
# Diccionario con parametros que se probaran durante la busqueda de hiperparámetros
# donde, 'msd' es (Mean Squared Difference), 'cosine' es (similitud coseno) y min_support es
# número mínimo de coincidencias (usuarios o ítems comunes) para considerar la similitud
param_grid = { 'sim_options' : {'name': ['msd','cosine'], \
                                'min_support': list(range(2,7)), \
                                'user_based': [False, True], \
                                'min_k':[5]}
             }

In [ ]:
# Creamos un objeto GridSearchCV para realizar la búsqueda de hiperparámetros
gridsearchKNNBaseline = GridSearchCV(KNNBaseline, param_grid, measures=['rmse'], \
                                      cv=2, n_jobs=-1) # cv: número de particiones para la validación cruzada (2 en este caso)
                                      # n_jobs=-1 usa todos los núcleos disponibles, es decir numero de trabajos en paralelo

gridsearchKNNBaseline.fit(data) # Entrenamiento del modelo

In [ ]:
# Obtener el mejor conjunto de hiperparámetros encontrados durante la búsqueda, especificamente los que lograron un mejor RMSE
gridsearchKNNBaseline.best_params["rmse"]
gridsearchKNNBaseline.best_score["rmse"]
gs_model=gridsearchKNNBaseline.best_estimator['rmse']

In [ ]:
trainset = data.build_full_trainset() ### esta función convierte todos los datos en entrnamiento, las funciones anteriores dividen  en entrenamiento y evaluación
model=gs_model.fit(trainset) ## se reentrena sobre todos los datos posibles (sin dividir)

Estimating biases using als...
Computing the msd similarity matrix...
Done computing similarity matrix.


In [ ]:
joblib.dump(model,'/content/drive/MyDrive/analitica 3/sistemas_recomendacion/salidas/modelo_colaborativo.joblib') # Guardamos el modelo para su posterior uso en el despliegue

['/content/drive/MyDrive/analitica 3/sistemas_recomendacion/salidas/modelo_colaborativo.joblib']

In [ ]:
### Hacer predicción con el modelo de recomendación
# uid: ID del usuario para quien se quiere predecir la calificación
# iid: ID del ítem (película) para el cual se quiere hacer la predicción
model.predict(uid=5, iid='5',r_ui='')

Prediction(uid=5, iid='5', r_ui='', est=3.5093755789717713, details={'was_impossible': False})

In [ ]:
#Crea el anti-testset: una lista de todas las combinaciones (usuario, película) que NO existen en el conjunto de entrenamiento
predset = trainset.build_anti_testset()
len(predset)

#Muestra las primeras 10 combinaciones usuario-película NO vistas
predset[0:10]

[(1, 318, 3.579442714350294),
 (1, 1704, 3.579442714350294),
 (1, 6874, 3.579442714350294),
 (1, 8798, 3.579442714350294),
 (1, 46970, 3.579442714350294),
 (1, 48516, 3.579442714350294),
 (1, 58559, 3.579442714350294),
 (1, 60756, 3.579442714350294),
 (1, 68157, 3.579442714350294),
 (1, 71535, 3.579442714350294)]

In [ ]:
# Utiliza el modelo entrenado (gs_model) para predecir las calificaciones
# de todas las combinaciones usuario-película que no han sido calificadas
predictions = gs_model.test(predset)

In [ ]:
#Muestra las primeras 10 predicciones
predictions[0:10]

[Prediction(uid=1, iid=318, r_ui=3.579442714350294, est=4.953888892997183, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid=1, iid=1704, r_ui=3.579442714350294, est=4.811858760280403, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid=1, iid=6874, r_ui=3.579442714350294, est=4.745184479485489, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid=1, iid=8798, r_ui=3.579442714350294, est=4.6102450867763745, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid=1, iid=46970, r_ui=3.579442714350294, est=4.139481333486526, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid=1, iid=48516, r_ui=3.579442714350294, est=4.9620836959599846, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid=1, iid=58559, r_ui=3.579442714350294, est=5, details={'actual_k': 40, 'was_impossible': False}),
 Prediction(uid=1, iid=60756, r_ui=3.579442714350294, est=4.232274503384751, details={'actual_k': 40, 'was_impossibl

**Total de combinaciones de peliculas no vistas por los usuarios llevadas a un dataframe** ↓

In [ ]:
predictions_df = pd.DataFrame(predictions) #Convertir la lista de predicciones en un dataframe
predictions_df.shape
predictions_df.head()
predictions_df['r_ui'].unique() ### promedio de ratings
predictions_df. sort_values(by='est',ascending=False) # Ordenar el dataframe enforma descendente respecto a est ---- est: Es la calificación predicha

uid     iid      r_ui       est  \
105295    53    6867  3.579443  5.000000   
104831    53     610  3.579443  5.000000   
104125    53    4014  3.579443  5.000000   
105549    53    3507  3.579443  5.000000   
550595   276  168252  3.579443  5.000000   
...      ...     ...       ...       ...   
5877       3    1381  3.579443  0.192631   
4680       3   63992  3.579443  0.170991   
881790   442    3593  3.579443  0.156023   
1192348  598     640  3.579443  0.088903   
1192338  598     540  3.579443  0.033542   

                                           details  
105295   {'actual_k': 17, 'was_impossible': False}  
104831   {'actual_k': 19, 'was_impossible': False}  
104125   {'actual_k': 19, 'was_impossible': False}  
105549   {'actual_k': 18, 'was_impossible': False}  
550595   {'actual_k': 36, 'was_impossible': False}  
...                                            ...  
5877     {'actual_k': 22, 'was_impossible': False}  
4680     {'actual_k': 16, 'was_impossible': False}  
881790   {'actual_k': 14, 'was_impossible': False}  
1192348   {'actual_k': 2, 'was_impossible': False}  
1192338   {'actual_k': 2, 'was_impossible': False}  

[1214174 rows x 5 columns]

In [ ]:
#Llevar dataframe de predicciones de peliculas a un csv
#predictions_df.to_csv('/content/drive/MyDrive/analitica 3/sistemas_recomendacion/salidas/predictions_movies.csv', index=False)

In [ ]:
#Funcion que genera recomendaciones personalizadas de películas para un usuario específico basadas en
#predicciones de ratings, y además incluye los títulos de las películas

def recomendaciones(user_id, n_recomend=10):

    # Filtrar las predicciones para el usuario y ordenar por la calificación estimada
    predictions_userID = predictions_df[predictions_df['uid'] == user_id].\
                        sort_values(by="est", ascending=False).head(n_recomend)

    # Seleccionar las columnas necesarias y renombrarlas
    recomendados = predictions_userID[['uid', 'iid', 'r_ui', 'est']]
    recomendados.columns = ['user_id', 'movie_id', 'promedio_rating_real', 'estimacion_rating']

    # Guardar las recomendaciones en la base de datos
    recomendados.to_sql('reco', conn, if_exists="replace", index=False)

    # Realizar la consulta SQL para obtener los títulos de las películas y eliminar duplicados
    recomendados = pd.read_sql('''SELECT a.*, b.title
                                  FROM reco a
                                  LEFT JOIN df_explorado b
                                  ON a.movie_id = b.movie_id''', conn)

    # Eliminar filas duplicadas
    recomendados = recomendados.drop_duplicates(subset=['movie_id', 'title'])

    return recomendados

# Ejemplo
df_recomendaciones=recomendaciones(user_id=100, n_recomend=5)
df_recomendaciones

user_id  movie_id  promedio_rating_real  estimacion_rating  \
0         100      3147              3.579443           4.407402   
222       100       318              3.579443           4.400224   
856       100     58559              3.579443           4.399551   
1452      100      2324              3.579443           4.392307   
1804      100     48516              3.579443           4.377912   

                                           title  
0                         Green Mile, The (1999)  
222             Shawshank Redemption, The (1994)  
856                      Dark Knight, The (2008)  
1452  Life Is Beautiful (La Vita è bella) (1997)  
1804                        Departed, The (2006)

Este sistema de recomendación sugiere películas a cada persona según los gustos de otros usuarios que tienen preferencias parecidas. En el resultado anterior podemos ver que, por ejemplo, al usuario 100 se le muestran películas que podrían gustarle, ordenadas por la calificación que se estima que les daría en base a calificaciones dadas por otros usarios parecidos que ya la vieron. Gracias a este método, se pueden recomendar películas que el usuario no ha visto, basándose en lo que le ha gustado a personas con gustos similares.

Este puede ser considerado un buen modelo y más en este caso que se tiene un base de datos bastante amplia ya que aprovecha los patrones de comportamiento de esta comunidad amplia, mejorando asi la precisión y utílidad de las recomendaciones de películas

### Uso de API para descripción de películas

- Obtener la Api Key de Gemini gratuita: https://aistudio.google.com/apikey (Tener cuenta previamente)

In [ ]:
#Creacion de lista de peliculas recomendadas al usuario x
lista_peliculas = df_recomendaciones.title.to_list()
lista_peliculas

['Green Mile, The (1999)',
 'Shawshank Redemption, The (1994)',
 'Dark Knight, The (2008)',
 'Life Is Beautiful (La Vita è bella) (1997)',
 'Departed, The (2006)']

In [ ]:
# Configuración de la clave API para acceder a los servicios de Gemini
genai.configure(api_key="AIzaSyCCppbwpZKqG-i5xocnFIJgVBUr_sBavew")

In [ ]:
# Crear una instancia del modelo generativo
# En este caso, se utiliza el modelo 'gemini-2.0-flash' que es una versión ligera y rápida de Gemini
lista_respuestas = []

model_ai = genai.GenerativeModel('gemini-2.0-flash')

# Definir una función para enviar un prompt y obtener la respuesta del modelo
def preguntar_a_gemini(prompt):
    # El método generate_content genera una respuesta para el prompt dado
    respuesta = model_ai.generate_content(prompt)
    # Se retorna solo el texto de la respuesta
    return respuesta.text


In [ ]:
#Promt para generar descripciones a las peliculas
for pelicula in lista_peliculas:
    promt = "Dame una descripción de 50 palabras de la siguiente pelicula: " + pelicula
    lista_respuestas.append(preguntar_a_gemini(promt))
lista_respuestas

['En la penitenciaría de Cold Mountain en 1935, el guardia Paul Edgecomb supervisa el bloque de la muerte. La vida de los guardias cambia cuando conocen a John Coffey, un enorme hombre negro con una discapacidad mental que es condenado por asesinar a dos niñas pequeñas. Sin embargo, pronto aprenden que Coffey posee un don milagroso que desafía su comprensión.\n',
 'Condenado injustamente por el asesinato de su esposa, Andy Dufresne pasa casi dos décadas en la prisión de Shawshank, donde se hace amigo de Red y encuentra la esperanza en actos de bondad aleatorios. Sin embargo, debajo de todo, Andy guarda un secreto.\n',
 'En "The Dark Knight", Batman se enfrenta al Joker, un agente del caos que sume a Gotham en la anarquía. El enfrentamiento pone a prueba los límites morales de Batman, mientras debe elegir entre el heroísmo y la justicia. La película es un thriller oscuro y psicológico lleno de acción y dilemas éticos.\n',
 'En la Italia de la Segunda Guerra Mundial, el encantador Guido,

In [ ]:
#Promt para averiguar el protagonista de cada una de las peliculas
lista_protagonista = []

for pelicula in lista_peliculas:
    promt = "Dime solo el nombre del protagonista principal de la pelicula: " + pelicula
    lista_protagonista.append(preguntar_a_gemini(promt))
lista_protagonista

['Paul Edgecomb\n',
 'Andy Dufresne\n',
 'Batman\n',
 'Guido\n',
 'William Costigan Jr.\n']

**Visualización del Dataframe completo con el protagonista y la descripción**

In [ ]:
#Agregar las dos columnas extras al dataframe de recomendaciones
df_recomendaciones['protagonista'] = lista_protagonista
df_recomendaciones['descripcion'] = lista_respuestas
df_recomendaciones

user_id  movie_id  promedio_rating_real  estimacion_rating  \
0         100      3147              3.579443           4.407402   
222       100       318              3.579443           4.400224   
856       100     58559              3.579443           4.399551   
1452      100      2324              3.579443           4.392307   
1804      100     48516              3.579443           4.377912   

                                           title            protagonista  \
0                         Green Mile, The (1999)         Paul Edgecomb\n   
222             Shawshank Redemption, The (1994)         Andy Dufresne\n   
856                      Dark Knight, The (2008)                Batman\n   
1452  Life Is Beautiful (La Vita è bella) (1997)                 Guido\n   
1804                        Departed, The (2006)  William Costigan Jr.\n   

                                            descripcion  
0     En la penitenciaría de Cold Mountain en 1935, ...  
222   Condenado injustamente por el asesinato de su ...  
856   En "The Dark Knight", Batman se enfrenta al Jo...  
1452  En la Italia de la Segunda Guerra Mundial, el ...  
1804  "The Departed" es un thriller intenso de Scors...